sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala

```diff
@@ -111,6 +113,24 @@ object DateTimeUtils {
     computedTimeZones.computeIfAbsent(timeZoneId, computeTimeZone)
   }
 
+  private val threadLocalComputedCalendarsMap =
+    new ThreadLocal[mutable.Map[TimeZone, (Calendar, Long)]] {
+      override def initialValue(): mutable.Map[TimeZone, (Calendar, Long)] = {
+        mutable.Map[TimeZone, (Calendar, Long)]()
+      }
+    }
+
+  def getCalendar(timeZone: TimeZone): Calendar = {
+    val (c, timeInMillis) = threadLocalComputedCalendarsMap.get()
+      .getOrElseUpdate(timeZone, {
+        val c = Calendar.getInstance(timeZone)
+        (c, c.getTimeInMillis)
+      })
+    c.clear()
+    c.setTimeInMillis(timeInMillis)
```

Comment:
I agree with @viirya 's comment. Do we need to set the value of `System.currentTimeMillis()`?

Reply:
@kiszk @viirya I've updated the code. Please review.